### PR TITLE
feat: Allow passing additional JWT claims to `TokenIssuer`

### PR DESCRIPTION
**Pull request by DavidZemon**
Signed-off-by: David Zemon <david@zemon.name>

## Hey, I just made a Pull Request!

We have, for some time now, been (ab)using the `ent` claim on the Backstage JWT to store credentials for the logged in user. Specifically, credentials to external systems that have not yet been integrated with our SSO provider. For example, we want the Backstage JWT to contain the end-user's BitBucket Personal Access Token, so we put that in the JWT as `ent: ['bbPat:xxxxx']`. This is.... not the ideal solution. But, the `TokenIssuer`, its `TokenParams` object, and even the specific implementation of `TokenFactory` prevent us from doing anything better. Even if we ignore the TypeScript compiler error and add additional claims on the input parameter, those claims will be ignored by `TokenFactory.issueToken()`.

So, I've added the ability to list any additional claims that a developer wants but ensured that those claims are the lowest priority when building the token so that a developer does not accidentally shoot themselves in the foot by overwriting important Backstage claims like `iss` and `exp`.

I've also updated the pertinent unit test to ensure that the overwriting is working correctly.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by benjdlambert** [comment]
Wondering if 1) we want to actually do this or if there's a better way, and 2) can we embed the types in there to get some typesafety or something rather than just `Record<string, string>`? @backstage/maintainers

**Comment by DavidZemon** [comment]
>  if 1) we want to actually do this or if there's a better way

I would love suggestions for a better way. We've been dealing with this for about a year now and I've yet to come up with one, but I'd be thrilled to hear of a better solution.

> 2) can we embed the types in there to get some typesafety or something rather than just `Record<string, string>`?

Maybe, but I'm not sure I see the value. What you definitely would _not_ want to do is hardcode additional properties from the list of registered public claims. My intention with this PR is to make it possible to add unregistered public claims as well as any private claims (as defined in sections [4.2 and 4.3 of the RFC](https://www.rfc-editor.org/rfc/rfc7519#section-4.2)).

So, you could define something like

```ts
interface TokenParams {
  sub: string;
  ent?: string[];
}
issueToken<T extends TokenParams>(params: T) {...}
```

but the implementation of `issueToken()` would remain the same. I don't see any value in this.
Sorry if I misunderstood something.

**Comment by DavidZemon** [comment]
@benjdlambert, does "@backstage/maintainers" actually do anything? Clicking that link leads me to a 404.
Is there someone specific that should be pinged on this?

**Comment by benjdlambert** [comment]
@DavidZemon yep it's our internal org team. Sorry for the delay, we've all been travelling to the US this week for a work thing, so we've been pretty inactive.

**Comment by benjdlambert** [comment]
Will pick this up next week!

**Comment by DavidZemon** [comment]
@benjdlambert , no problem, that explains a bit (I was also told this earlier today on discord).
Did you close this pr intentionally?

**Comment by benjdlambert** [comment]
OH NOPE.

File: plugins/auth-backend/api-report.md [hunk 1]
```diff
@@ -706,7 +706,7 @@ export type TokenParams = {
   claims: {
     sub: string;
     ent?: string[];
-  };
```
Comment (Rugvip) [inline plugins/auth-backend/api-report.md:709]: I'm thinking this should probably be `Record<string, JsonValue>`?

File: plugins/auth-backend/src/identity/types.ts [hunk 2]
```diff
@@ -28,13 +28,19 @@ export interface AnyJWK extends Record<string, string> {
  * @public
  */
 export type TokenParams = {
-  /** The claims that will be embedded within the token */
+  /**
+   * The claims that will be embedded within the token. At a minimum, this should include
+   * the subject claim, `sub`. It is common to also list entity ownership relations in the
+   * `ent` list. Additional claims may also be added at the developer's discretion except
+   * for the following list, which will be overwritten by the TokenIssuer: `iss`, `aud`,
+   * `iat`, and `exp`.
```
Comment (Rugvip) [inline plugins/auth-backend/src/identity/types.ts:36]: Could we add a note here that any additional claims are always potentially breaking? basically we reserve the right to add new claims in the future without calling it a breaking change.

**Review by Rugvip — CHANGES_REQUESTED** [review]
👍 , two minor things

File: plugins/auth-backend/src/identity/types.ts (hunk 2 above)
Comment (DavidZemon) [reply plugins/auth-backend/src/identity/types.ts:36]: Absolutely. Should that be phrased "Backstage maintains...", "The Backstage authors maintain...", or "Spotify maintains..."?

File: plugins/auth-backend/api-report.md (hunk 1 above)
Comment (DavidZemon) [reply plugins/auth-backend/api-report.md:709]: Great point.

Comment (DavidZemon) [reply plugins/auth-backend/api-report.md:709]: Done

**Review by benjdlambert — APPROVED** [review]
Nice! Thanks for this! :pray:

**Comment by pietervogelaar** [comment]
Thanks so much for this! This is exactly what I needed to integrate with an external system.

**Pull request merged**